File: emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLEMathAccel.java
// -*- Mode: Java; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// Multi-Phasic Applications: SquirrelJME
//     Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the Mozilla Public License Version 2.0.
// See license.mkd for licensing and copyright information.
// ---------------------------------------------------------------------------

package cc.squirreljme.vm.springcoat;

import cc.squirreljme.jvm.mle.MathAccelShelf;
import cc.squirreljme.jvm.mle.MathShelf;

/**
 * Functions for {@link MLEMathAccel}.
 *
 * @since 2025/05/03
 */
public enum MLEMathAccel
	implements MLEFunction
{
	/** {@link MathAccelShelf#accel()}. */
	ACCEL("accel:()I")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.accel();
		}
	},
	
	/** {@link MathAccelShelf#acos(double)}. */
	ACOS("acos:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.acos((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#asin(double)}. */
	ASIN("asin:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.asin((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#atan(double)}. */
	ATAN("atan:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.atan((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#atan2(double, double)}. */
	ATAN_2("atan2:(DD)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.atan2((double)__args[0],
				(double)__args[1]);
		}
	},
	
	/** {@link MathAccelShelf#ceil(double)}. */
	CEIL("ceil:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.ceil((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#cos(double)}. */
	COS("cos:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.cos((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#exp(double)}. */
	EXP("exp:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.exp((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#floor(double)}. */
	FLOOR("floor:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.floor((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#log(double)}. */
	LOG("log:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.log((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#pow(double, double)}. */
	POW("pow:(DD)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.pow((double)__args[0],
				(double)__args[1]);
		}
	},
	
	/** {@link MathAccelShelf#round(double)}. */
	ROUND("round:(D)J")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.round((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#signum(double)}. */
	SIGNUM("signum:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.signum((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#sin(double)}. */
	SIN("sin:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.sin((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#sqrt(double)}. */
	SQRT("sqrt:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.sqrt((double)__args[0]);
		}
	},
	
	/** {@link MathAccelShelf#tan(double)}. */
	TAN("tan:(D)D")
	{
		/**
		 * {@inheritDoc}
		 * @since 2025/05/03
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			return MathAccelShelf.tan((double)__args[0]);
		}
	},
	
	/* End. */
	;
	
	/** The dispatch key. */
	protected final String key;
	
	/**
	 * Initializes the dispatcher info.
	 *
	 * @param __key The key.
	 * @throws NullPointerException On null arguments.
	 * @since 2020/06/18
	 */
	MLEMathAccel(String __key)
		throws NullPointerException
	{
		if (__key == null)
			throw new NullPointerException("NARG");
		
		this.key = __key;
	}
	
	/**
	 * {@inheritDoc}
	 * @since 2020/06/18
	 */
	@Override
	public String key()
	{
		return this.key;
	}
}

File: emulators/emulator-base-native/c/mle_mathaccel.c
/* -*- Mode: C++; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// SquirrelJME
//	 Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the Mozilla Public License Version 2.0.
// See license.mkd for licensing and copyright information.
// -------------------------------------------------------------------------*/

#include <stdlib.h>

#include "squirreljme.h"

#define FORWARD_CLASS "cc/squirreljme/jvm/mle/MathAccelShelf"
#define FORWARD_CLASS_NAME MathAccel
#define FORWARD_NATIVE_CLASS "cc/squirreljme/emulator/EmulatedMathAccelShelf"

#define FORWARD_DESC_accel \
	DESC_METHOD(DESC_INT, )
#define FORWARD_DESC_acos \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_asin \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_atan \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_atan2 \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE DESC_DOUBLE)
#define FORWARD_DESC_ceil \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_cos \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_exp \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_floor \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_log \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_pow \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE DESC_DOUBLE)
#define FORWARD_DESC_round \
	DESC_METHOD(DESC_LONG, DESC_DOUBLE)
#define FORWARD_DESC_signum \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_sin \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_sqrt \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)
#define FORWARD_DESC_tan \
	DESC_METHOD(DESC_DOUBLE, DESC_DOUBLE)

FORWARD_IMPL(MathAccel, accel,
	jint, Integer,
	FORWARD_IMPL_none(),
	FORWARD_IMPL_none())
FORWARD_IMPL(MathAccel, acos,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, asin,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, atan,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, atan2,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble a, jdouble b),
	FORWARD_IMPL_pass(a, b))
FORWARD_IMPL(MathAccel, ceil,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, cos,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, exp,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, floor,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, log,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, pow,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble a, jdouble b),
	FORWARD_IMPL_pass(a, b))
FORWARD_IMPL(MathAccel, round,
	jlong, Long,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, signum,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, sin,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, sqrt,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))
FORWARD_IMPL(MathAccel, tan,
	jdouble, Double,
	FORWARD_IMPL_args(jdouble value),
	FORWARD_IMPL_pass(value))

static const JNINativeMethod mleMathAccelMethods[] =
{
	FORWARD_list(MathAccel, accel),
	FORWARD_list(MathAccel, acos),
	FORWARD_list(MathAccel, asin),
	FORWARD_list(MathAccel, atan),
	FORWARD_list(MathAccel, atan2),
	FORWARD_list(MathAccel, ceil),
	FORWARD_list(MathAccel, cos),
	FORWARD_list(MathAccel, exp),
	FORWARD_list(MathAccel, floor),
	FORWARD_list(MathAccel, log),
	FORWARD_list(MathAccel, pow),
	FORWARD_list(MathAccel, round),
	FORWARD_list(MathAccel, signum),
	FORWARD_list(MathAccel, sin),
	FORWARD_list(MathAccel, sqrt),
	FORWARD_list(MathAccel, tan),
};

jint JNICALL mleMathAccelInit(JNIEnv* env, jclass classy)
{
	return (*env)->RegisterNatives(env,
		(*env)->FindClass(env, "cc/squirreljme/jvm/mle/MathAccelShelf"),
		mleMathAccelMethods, sizeof(mleMathAccelMethods) /
			sizeof(JNINativeMethod));
}
